### PR TITLE
Mono publisher and consumer removal

### PR DESCRIPTION
**Pull request by CarlosGamero**
Multiple consumers and publishers can accomplish the same tasks as single ones, but they add layer of complexity by requiring features to be implemented in both. As a result, we have decided to remove the single ones to enhance maintainability.

Note that this update will impact all packages. Thank you for your understanding.

TODO -> Adjust documentation once PR is reviewed ✅

### CODE REVIEW

File: packages/amqp/lib/AbstractAmqpPublisher.ts [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+import type { AMQPLocator, AMQPCreationConfig, AMQPDependencies } from './AbstractAmqpService'
+import { AbstractAmqpService } from './AbstractAmqpService'
+
+export type AMQPPublisherOptions<MessagePayloadType extends object> = QueuePublisherOptions<
+  AMQPCreationConfig,
+  AMQPLocator,
+  MessagePayloadType
+>
+
+export abstract class AbstractAmqpPublisher<MessagePayloadType extends object>
```
Comment (CarlosGamero) [inline packages/amqp/lib/AbstractAmqpPublisher.ts:21]: This is the old `AbstractAmqpBasePublisher` combined with `AbstractAmqpPublisherMultiSchema`

File: packages/amqp/lib/AbstractAmqpConsumer.ts [hunk 2]
```diff
@@ -1,77 +1,151 @@
...
+> = QueueConsumerOptions<
+  AMQPCreationConfig,
+  AMQPLocator,
+  MessagePayloadType,
+  ExecutionContext,
+  PrehandlerOutput
+>
 
-export abstract class AbstractAmqpBaseConsumer<
+export abstract class AbstractAmqpConsumer<
```
Comment (CarlosGamero) [inline packages/amqp/lib/AbstractAmqpConsumer.ts:40]: Combined with `AbstractAmqpConsumerMultiSchema`

File: packages/core/lib/types/queueOptionsTypes.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+  locatorConfig: QueueLocatorType
+  deletionConfig?: DeletionConfig
+  creationConfig?: never
+} & CommonQueueOptions
+
+export type QueueOptions<
+  CreationConfigType extends CommonCreationConfigType,
+  QueueLocatorType extends object,
+> = CommonQueueOptions &
+  (NewQueueOptions<CreationConfigType> | ExistingQueueOptions<QueueLocatorType>)
```
Comment (CarlosGamero) [inline packages/core/lib/types/queueOptionsTypes.ts:52]: Combining `NewQueueOptions` and `ExistingQueueOptions`, we were doing it manually on all packages so with this we simplify types there

File: packages/core/lib/types/queueOptionsTypes.ts [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+export type QueueConsumerOptions<
+  CreationConfigType extends object,
+  QueueLocatorType extends object,
+  MessagePayloadSchemas extends object,
+  ExecutionContext,
+  PrehandlerOutput = undefined,
+> = QueueOptions<CreationConfigType, QueueLocatorType> & {
+  handlers: MessageHandlerConfig<MessagePayloadSchemas, ExecutionContext, PrehandlerOutput>[]
+}
```
Comment (CarlosGamero) [inline packages/core/lib/types/queueOptionsTypes.ts:70]: Common types for publishers and consumers with common options across packages

File: packages/sqs/lib/sqs/AbstractSqsPublisher.ts [hunk 5]
```diff
@@ -48,7 +50,37 @@ export abstract class AbstractSqsPublisherMultiSchema<MessagePayloadType extends
       throw messageSchemaResult.error
     }
 
-    return this.internalPublish(message, messageSchemaResult.result, options)
```
Comment (CarlosGamero) [inline packages/sqs/lib/sqs/AbstractSqsPublisher.ts:51]: Moving internal publish to publisher and it is a publisher only method

File: packages/sqs/lib/sqs/AbstractSqsConsumer.ts [hunk 6]
```diff
@@ -29,33 +33,44 @@ export type ExtraSQSCreationParams = {
...
+  CreationConfigType,
+  QueueLocatorType,
+  MessagePayloadSchemas,
+  ExecutionContext,
+  PrehandlerOutput
+> & {
   consumerOverrides?: Partial<ConsumerOptions>
 }
-
 export abstract class AbstractSqsConsumer<
```
Comment (CarlosGamero) [inline packages/sqs/lib/sqs/AbstractSqsConsumer.ts:55]: Combined with `AbstractSqsConsumerMultiSchema`

File: packages/sqs/lib/sqs/AbstractSqsService.ts [hunk 7]
```diff
@@ -48,79 +40,34 @@ export abstract class AbstractSqsService<
...
+  ExecutionContext
 > {
   protected readonly sqsClient: SQSClient
+
   // @ts-ignore
-  public queueUrl: string
+  protected queueName: string
   // @ts-ignore
-  public queueName: string
+  protected queueUrl: string
```
Comment (CarlosGamero) [inline packages/sqs/lib/sqs/AbstractSqsService.ts:50]: I guess we don't want to expose those properties so marking them as protected

**Review by kibertoad — APPROVED** [review]
Impressive!

**Comment by kibertoad** [comment]
@CarlosGamero Can you add "UPGRADING.md" file with explanation for the migration path?

**Comment by CarlosGamero** [comment]
@kibertoad Of course, I will add it and address the required changes on readme 🙇

File: UPGRADING.md [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+
+### Migration Steps
+#### Multi consumers and publishers
+If you are using the multi consumer or consumer, you will only need to rename the class you are extending, and it should 
+work as before.
+- `AbstractSqsMultiConsumer` -> `AbstractSqsConsumer`
+- `AbstractSqsMultiPublisher` -> `AbstractSqsPublisher`
+
+#### Mono consumers and publishers
+If you are using the mono consumer or publisher, they no longer exists, so you will need to transform your code to use
```
Comment (kibertoad) [inline UPGRADING.md:21]:
```suggestion
If you are using the mono consumer or publisher, they no longer exist, so you will need to adjust your code to use
```

File: UPGRADING.md [hunk 9]
```diff
@@ -0,0 +1,146 @@
+# Upgrading Guide
+
+We have introduced the following breaking changes on version `12.0.0`, please follow the steps below to update your code
+from the previous version to the new one.
+
+## Breaking Changes
+
+### Description of Breaking Change
+Multiple consumers and publishers can accomplish the same tasks as single ones, but they add layer of complexity by
```
Comment (kibertoad) [inline UPGRADING.md:9]:
```suggestion
Multi consumers and publishers can accomplish the same tasks as mono ones, but they add extra layer of complexity by 
```

File: UPGRADING.md [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+
+We have introduced the following breaking changes on version `12.0.0`, please follow the steps below to update your code
+from the previous version to the new one.
+
+## Breaking Changes
+
+### Description of Breaking Change
+Multiple consumers and publishers can accomplish the same tasks as single ones, but they add layer of complexity by 
+requiring features to be implemented in both.
+As a result, we have decided to remove the single ones to enhance maintainability.
```
Comment (kibertoad) [inline UPGRADING.md:11]:
```suggestion
As a result, we have decided to remove the mono ones to enhance maintainability.
```

File: UPGRADING.md (hunk 8 above)
Comment (CarlosGamero) [reply UPGRADING.md]: Fixed, thanks Igor 🙇

File: UPGRADING.md (hunk 9 above)
Comment (CarlosGamero) [reply UPGRADING.md]: Fixed, thanks 🥇

File: UPGRADING.md (hunk 10 above)
Comment (CarlosGamero) [reply UPGRADING.md]: Fixed! thanks 💯

**Comment by CarlosGamero** [comment]
@kibertoad re-requesting review with the doc changes + upgrading guide but please no rush on reviewing and enjoy the Sunday 🙏 

Btw, sorry for the huge PR ... 😓

**Review by kibertoad — APPROVED** [review]
Great success!

**Comment by kibertoad** [comment]
@CarlosGamero shall we merge already?

**Comment by CarlosGamero** [comment]
I wont  be able to release it today, so I was thinking on merging it tomorrow and do the release just after

**Pull request merged**